File: content/tracks/dsa/problems/lc-0347-top-k-frequent-elements/solution.py
from typing import Dict, List


class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        counts: Dict[int, int] = {}
        for num in nums:
            counts[num] = counts.get(num, 0) + 1
        # buckets[f]: the values that appear exactly f times (f is at most len(nums))
        buckets: List[List[int]] = [[] for _ in range(len(nums) + 1)]
        for num, count in counts.items():
            buckets[count].append(num)
        result: List[int] = []
        for count in range(len(nums), 0, -1):  # most frequent first
            for num in buckets[count]:
                result.append(num)
                if len(result) == k:
                    return result
        return result

File: content/tracks/dsa/problems/lc-0347-top-k-frequent-elements/solution.go
package main

func topKFrequent(nums []int, k int) []int {
	counts := map[int]int{}
	for _, num := range nums {
		counts[num]++
	}
	// buckets[f]: the values that appear exactly f times (f is at most len(nums))
	buckets := make([][]int, len(nums)+1)
	for num, count := range counts {
		buckets[count] = append(buckets[count], num)
	}
	result := make([]int, 0, k)
	for count := len(nums); count > 0; count-- { // most frequent first
		for _, num := range buckets[count] {
			result = append(result, num)
			if len(result) == k {
				return result
			}
		}
	}
	return result
}
